### ISSUE TITLE
Демонстрационный пример

### ISSUE THREAD

**Issue by igsha** [13Fire13nyu/MasterDegree#4]
* Создать отдельный ipynb-файл, назвать demo или как-то похоже
* Открыть в demo подготовленный файл из #3 
* Открыть в demo натренированную модель в формате SavedModel (пример сохранения и загрузки https://www.tensorflow.org/api_docs/python/tf/keras/models/save_model#usage_2)
* Разбить демострационный файл на блоки, которые потом подавать на вход модели
* Модель возращает ноту в некотором блоке. Объединять блоки по времени
* Сравнить ноты и их временные отметки с временными метками из #3 
* Вывести метрику качества ([precision](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Precision) и [recall](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Recall))

**Comment by 13Fire13nyu** [comment]
Ещё у меня есть вопросы по поводу деления файла на блоки. Как минимум, какой должен быть размер блока, как его выбирать? И что значить "объединять блоки по времени"?

**Comment by igsha** [comment]
Размер блока выбирается из размерности загруженной модели. Это должно быть сделано автоматом.
Объединять результаты выхода модели по времени. Модель же возвращает только одну точку. По идее нужно построить график полученных нот по времени с тем же графиком, но полученным из csv-файла.

**Comment by 13Fire13nyu** [comment]
> Размер блока выбирается из размерности загруженной модели. Это должно быть сделано автоматом. Объединять результаты выхода модели по времени. Модель же возвращает только одну точку. По идее нужно построить график полученных нот по времени с тем же графиком, но полученным из csv-файла.

Размерность модели можно как-то считать из нее?

**Comment by igsha** [comment]
Модель состоит из тензоров. Тензоры имеют фиксированные размерности. Поэтому данные для модели могут быть поданы только определённой размерности. Исключение составляет количество образцов данных. Один образец - один результат прогноза. N образцов - N результатов прогноза.
> Или рассматриваем куски с размером, равным размеру аудио, подаваемых ранее для обучения?

Да, можно и так делать.

**Comment by 13Fire13nyu** [comment]
> 

Я же могу, по идее, просто получить спектрограмму куска звукового сигнала и подать ее модели для предсказания?

**Comment by igsha** [comment]
Да, можете. Просто начните делать. Присылайте сюда конкретные ошибки или результаты, которые получаете.

**Comment by 13Fire13nyu** [comment]
> Да, можете. Просто начните делать. Присылайте сюда конкретные ошибки или результаты, которые получаете.

Ну, собственно, я потому и спрашиваю, потому что вот:
![image](https://user-images.githubusercontent.com/47567199/164984443-198f6738-ac40-4571-99c7-0e8cf757c634.png)

**Comment by igsha** [comment]
В сообщении говорится, что входной тензор должен иметь размерность 806x257. Это точно та же модель, что была в отчёте?

**Comment by 13Fire13nyu** [comment]
> В сообщении говорится, что входной тензор должен иметь размерность 806x257. Это точно та же модель, что была в отчёте?

Ну я её сохраняла.

**Comment by igsha** [comment]
Почитайте тогда документацию. Возможно `SavedModel` возвращает объект, а не функцию.

**Comment by 13Fire13nyu** [comment]
> Почитайте тогда документацию. Возможно `SavedModel` возвращает объект, а не функцию.

Как я поняла, SavedModel экспортирует объект в формат SavedModel

**Comment by 13Fire13nyu** [comment]
> Почитайте тогда документацию. Возможно `SavedModel` возвращает объект, а не функцию.

Есть еще один способ сохранить модель: model.save(r"D:\magistracy\diplom\model\model.h5"):
![image](https://user-images.githubusercontent.com/47567199/164987382-cc80fd66-6408-49df-a4dc-15ec9a7a3b00.png)

Потом она загружается так:
from tensorflow.keras.models import load_model
model = load_model(r"D:\magistracy\diplom\model\model.h5")

Но у меня все равно вылазит ошибка:
![image](https://user-images.githubusercontent.com/47567199/164987361-e488cc54-c643-4f9f-9718-1069001b9706.png)

**Comment by igsha** [comment]
У Вас на вход модели подаётся 806x257x1, а должно подаваться как минимум 1x806x257x1. Предсказывание в модели всегда работает на пачку входных данных, а не на единичный экзепляр. Надо либо нарезать весь аудиофайл, либо делать `tf.expand_dims(spectrogram, axis=0)` (да, ещё раз).

**Comment by 13Fire13nyu** [comment]
Пока у меня на выходе следующие результаты: 
![image](https://user-images.githubusercontent.com/47567199/165127745-92308eba-e98d-48a6-a700-73d74737ca09.png)
Я разбивала аудио на блоки по 16000 отсчетов. Первое значение - время начала в секундах, второе значение - время окончания в секундах, третье - нота.

**Comment by igsha** [comment]
> Сравнить ноты и их временные отметки с временными метками из Файл для демонстраций #3

Постройте ступенчатые графики, чтобы было наглядно видно, где мы ошибаемся. По сути нотный стан - это тоже график.

>  но я думаю, что блоки надо делать меньше, потому что попадает несколько нот в один отрывок.

Эта проблема будет решена при выборе размера окна. Сейчас можно об этом не беспокоиться.

> Вывести метрику качества (precision и recall)

В этом задании ещё это осталось сделать. Ничего улучшать не нужно. Нам нужно только подготовить инфраструктуру.

**Comment by 13Fire13nyu** [comment]
> > Сравнить ноты и их временные отметки с временными метками из Файл для демонстраций #3
> 
> Постройте ступенчатые графики, чтобы было наглядно видно, где мы ошибаемся. По сути нотный стан - это тоже график.
> 
> > но я думаю, что блоки надо делать меньше, потому что попадает несколько нот в один отрывок.
> 
> Эта проблема будет решена при выборе размера окна. Сейчас можно об этом не беспокоиться.
> 
> > Вывести метрику качества (precision и recall)
> 
> В этом задании ещё это осталось сделать. Ничего улучшать не нужно. Нам нужно только подготовить инфраструктуру.

Ступенчатый график - это где по абсциссе отложено время, а по ординате - нота? И разве можно считать точность для результатов разной размерности? Потому что количество результатов не совпадает с количеством строк в csv-файле.

**Comment by 13Fire13nyu** [comment]
> Про ступенчатый график не буду отвечать. Это гуглится. Про разную размерность не понял. Всегда разные графики рисовал на одном плоте - тут не бывает проблем. Результаты и не должны совпадать с количеством строк в csv-файле, потому что у каждого графика свои `x` и свои `y`.

В размерности я имела в виду, что надо же как-то именно сравнивать полученное значение с реальным. Обычно, насколько я помню, y_pred и y_true были одинаковой длины. Здесь же такого не наблюдается.

**Comment by 13Fire13nyu** [comment]
Вот результаты со ступенчатым графиком.
![image](https://user-images.githubusercontent.com/47567199/165341874-32d1a221-5430-42b8-98d6-2703c5e3765f.png)
Но с подсчетом precision и recall все еще есть проблемы, потому что размер y_true и y_pred не совпадает.

**Comment by igsha** [comment]
Ну понятно, что Precision и Recall видимо вычисляют только бинарные классы. Погуглите, может tf не умеет для мультиклассов метрики вычислять. Тогда используйте обычное `accuracy`.

Не понимаю Ваших графиков. Судя по оригиналу, в начале мелодии идёт `C->D->E->D->C`, в `y_pred` получилось `C->D->D->E->D->C#`, что очень похоже на оригинал. Но графики получились абсолютно разными, как будто неправильно выставлены метрики.

Использовать `np.linspace` не нужно. Можно просто ориентироваться на первую секунду ноты:
```python
timescale1 = wf_notes[:][0]
timescale2 = real_notes[:][0]
```

**Comment by 13Fire13nyu** [comment]
> Ну понятно, что Precision и Recall видимо вычисляют только бинарные классы. Погуглите, может tf не умеет для мультиклассов метрики вычислять. Тогда используйте обычное `accuracy`.
> 
> Не понимаю Ваших графиков. Судя по оригиналу, в начале мелодии идёт `C->D->E->D->C`, в `y_pred` получилось `C->D->D->E->D->C#`, что очень похоже на оригинал. Но графики получились абсолютно разными, как будто неправильно выставлены метрики.
> 
> Использовать `np.linspace` не нужно. Можно просто ориентироваться на первую секунду ноты:
> 
> ```python
> timescale1 = wf_notes[:][0]
> timescale2 = real_notes[:][0]
> ```

Да, увидела, что precision и recall в tf не поддерживают мультиклассовость. Остальное посмотрю.

**Comment by 13Fire13nyu** [comment]
> > Ну понятно, что Precision и Recall видимо вычисляют только бинарные классы. Погуглите, может tf не умеет для мультиклассов метрики вычислять. Тогда используйте обычное `accuracy`.
> > Не понимаю Ваших графиков. Судя по оригиналу, в начале мелодии идёт `C->D->E->D->C`, в `y_pred` получилось `C->D->D->E->D->C#`, что очень похоже на оригинал. Но графики получились абсолютно разными, как будто неправильно выставлены метрики.
> > Использовать `np.linspace` не нужно. Можно просто ориентироваться на первую секунду ноты:
> > ```python
> > timescale1 = wf_notes[:][0]
> > timescale2 = real_notes[:][0]
> > ```
> 
> Да, увидела, что precision и recall в tf не поддерживают мультиклассовость. 
По поводу значений, которые выводит сеть. До этого я показывала результаты при шаге 16000 отсчётов, и там как раз было `C->D->D->E->D->C#`, однако сейчас уже другой шаг, который установлен в соответствии с количеством строк в csv-файле для возможности сравнения y_true и y_pred. Сейчас я не могу привести, какие там ноты получились, но вечером уже посмотрю, что там. 
Остальное посмотрю.

**Comment by 13Fire13nyu** [comment]
> > Ну понятно, что Precision и Recall видимо вычисляют только бинарные классы. Погуглите, может tf не умеет для мультиклассов метрики вычислять. Тогда используйте обычное `accuracy`.
> > Не понимаю Ваших графиков. Судя по оригиналу, в начале мелодии идёт `C->D->E->D->C`, в `y_pred` получилось `C->D->D->E->D->C#`, что очень похоже на оригинал. Но графики получились абсолютно разными, как будто неправильно выставлены метрики.
> > Использовать `np.linspace` не нужно. Можно просто ориентироваться на первую секунду ноты:
> > ```python
> > timescale1 = wf_notes[:][0]
> > timescale2 = real_notes[:][0]
> > ```
> 
> Да, увидела, что precision и recall в tf не поддерживают мультиклассовость. Остальное посмотрю.

Вот реально определяемые ноты:
![image](https://user-images.githubusercontent.com/47567199/165597715-3d3b45c2-632d-46c6-9511-1771f0641496.png)
Также по поводу linspace: я попробовала сделать так, как предложили Вы, но у меня вывелась ошибка несоответствия размерностей:
![image](https://user-images.githubusercontent.com/47567199/165597904-418756d1-a99c-4bb2-8ef3-5e4d0b1fce6d.png)
По поводу метрик: я посчитала accuracy, а еще использовала sklearn для подсчета precision и recall:
![image](https://user-images.githubusercontent.com/47567199/165599499-8a1c2395-db10-479c-b8b3-014d6f3f5767.png)
![image](https://user-images.githubusercontent.com/47567199/165599536-09eff494-bf18-47e4-81cc-0aa9babee0f2.png)

**Comment by 13Fire13nyu** [comment]
Дополнительно вывела еще отчет о классификации:
![image](https://user-images.githubusercontent.com/47567199/165603088-d815cd0d-065f-4010-bc87-199734ec342e.png)
Видимо, такое деление на блоки дает очень плохие результаты.

**Comment by igsha** [comment]
> По поводу метрик: я посчитала accuracy, а еще использовала sklearn для подсчета precision и recall:

Осталвяем такой вариант подсчёта метрик.

> Также по поводу linspace: я попробовала сделать так, как предложили Вы, но у меня вывелась ошибка несоответствия размерностей:

Это гуглится. Очевидно, что нужно `:` и `0` написать как-то иначе. Всё зависит от типа массива.

> Вот реально определяемые ноты:

Похоже, что неправильно заданы ноты. При обучении у Вас было другое соответствие нот. Соответствие нот и индекса класса должно быть точно таким, как это было сделано при обучении. Есть также вероятность, что сеть обучалась по неверному соответствию меток класса и индекса класса.

Выложите код на гитхаб, чтобы можно было посмотреть, где ошибки, а не гадать.

**Comment by 13Fire13nyu** [comment]
> > По поводу метрик: я посчитала accuracy, а еще использовала sklearn для подсчета precision и recall:
> 
> Осталвяем такой вариант подсчёта метрик.
> 
> > Также по поводу linspace: я попробовала сделать так, как предложили Вы, но у меня вывелась ошибка несоответствия размерностей:
> 
> Это гуглится. Очевидно, что нужно `:` и `0` написать как-то иначе. Всё зависит от типа массива.
> 
> > Вот реально определяемые ноты:
> 
> Похоже, что неправильно заданы ноты. При обучении у Вас было другое соответствие нот. Соответствие нот и индекса класса должно быть точно таким, как это было сделано при обучении. Есть также вероятность, что сеть обучалась по неверному соответствию меток класса и индекса класса.
> 
> Выложите код на гитхаб, чтобы можно было посмотреть, где ошибки, а не гадать.
Выложила код в гитхаб.

**Comment by 13Fire13nyu** [comment]
> > > По поводу метрик: я посчитала accuracy, а еще использовала sklearn для подсчета precision и recall:
> > 
> > 
> > Осталвяем такой вариант подсчёта метрик.
> > > Также по поводу linspace: я попробовала сделать так, как предложили Вы, но у меня вывелась ошибка несоответствия размерностей:
> > 
> > 
> > Это гуглится. Очевидно, что нужно `:` и `0` написать как-то иначе. Всё зависит от типа массива.
> > > Вот реально определяемые ноты:
> > 
> > 
> > Похоже, что неправильно заданы ноты. При обучении у Вас было другое соответствие нот. Соответствие нот и индекса класса должно быть точно таким, как это было сделано при обучении. Есть также вероятность, что сеть обучалась по неверному соответствию меток класса и индекса класса.
> > Выложите код на гитхаб, чтобы можно было посмотреть, где ошибки, а не гадать.

Выложила код в гитхаб.

**Comment by 13Fire13nyu** [comment]
> > > > По поводу метрик: я посчитала accuracy, а еще использовала sklearn для подсчета precision и recall:
> > > 
> > > 
> > > Осталвяем такой вариант подсчёта метрик.
> > > > Также по поводу linspace: я попробовала сделать так, как предложили Вы, но у меня вывелась ошибка несоответствия размерностей:
> > > 
> > > 
> > > Это гуглится. Очевидно, что нужно `:` и `0` написать как-то иначе. Всё зависит от типа массива.
> > > > Вот реально определяемые ноты:
> > > 
> > > 
> > > Похоже, что неправильно заданы ноты. При обучении у Вас было другое соответствие нот. Соответствие нот и индекса класса должно быть точно таким, как это было сделано при обучении. Есть также вероятность, что сеть обучалась по неверному соответствию меток класса и индекса класса.
> > > Выложите код на гитхаб, чтобы можно было посмотреть, где ошибки, а не гадать.
> 
> Выложила код в гитхаб. Я поняла, почему  так плохо прогнозировались ноты: дело оказалось в их нумерации. Ну я, вроде, этот момент подкорректировала.

**Comment by igsha** [comment]
Уже хорошо. Будем считать, что демо-пример готов к использованию. В будущем будем периодически возвращаться к нему, чтобы дополнять.

**Comment by 13Fire13nyu** [comment]
[Код](https://github.com/13Fire13nyu/MasterDegree/blob/master/demo_v1.ipynb)
